File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/com/ririnto/sinon/harness/rules/text/SkillFrontmatterRule.java
package com.ririnto.sinon.harness.rules.text;

import com.ririnto.sinon.harness.rules.HarnessCheckHelper;
import com.ririnto.sinon.harness.rules.HarnessCheckRule;
import com.ririnto.sinon.harness.core.RuleContext;
import com.ririnto.sinon.harness.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Stream;

/**
 * Rule that requires skill files to have proper frontmatter.
 */
public enum SkillFrontmatterRule implements HarnessCheckRule {
    INSTANCE;

    private static final String CATEGORY = "skillFrontmatter";

    @Override
    public String category() {
        return "skillFrontmatter";
    }

    @Override
    public boolean applies(RuleContext ctx) {
        return ctx.manifest().isEnabled(CATEGORY);
    }

    @Override
    public Collection<Finding> validate(RuleContext ctx) throws MojoExecutionException {
        final Path root = ctx.root();
        final JsonNode manifest = ctx.manifest().raw();
        final Path directory = root.resolve(".claude/skills");
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        if (!HarnessCheckHelper.isSafeDirectory(root, directory)) {
            return List.of(Finding.of(severity, CATEGORY, ".claude/skills must contain at least one SKILL.md"));
        }
        final List<Path> skillFiles = HarnessCheckHelper.safeFileOrWalk(root, directory).stream()
                .filter(f -> f.getFileName().toString().equals("SKILL.md"))
                .toList();
        if (skillFiles.isEmpty()) {
            return List.of(Finding.of(severity, CATEGORY, ".claude/skills must contain at least one SKILL.md"));
        }
        final Pattern descPattern = Pattern.compile("(?m)^description:\\s*.+$");
        return skillFiles.stream()
                .flatMap(f -> {
                    try {
                        return validateSkillFile(root, f, descPattern, severity).stream();
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }

    private List<Finding> validateSkillFile(Path root, Path file, Pattern descPattern, String severity) throws MojoExecutionException {
        final String text = HarnessCheckHelper.readFile(root, file);
        final String relative = root.relativize(file).toString();
        return Stream.<Stream<Finding>>of(
                text.startsWith("---") ? Stream.<Finding>empty() : Stream.of(Finding.of(severity, CATEGORY, "skill missing frontmatter: " + relative)),
                descPattern.matcher(text).find() ? Stream.<Finding>empty() : Stream.of(Finding.of(severity, CATEGORY, "skill missing description: " + relative))
        ).flatMap(s -> s).toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/com/ririnto/sinon/harness/rules/text/SkillFrontmatterRule.kt
package com.ririnto.sinon.harness.rules.text

import com.ririnto.sinon.harness.ast.HarnessAstResults.Finding
import com.ririnto.sinon.harness.core.JsonAccess
import com.ririnto.sinon.harness.core.RuleContext
import com.ririnto.sinon.harness.rules.HarnessCheckRule
import kotlinx.serialization.json.jsonObject
import org.commonmark.ext.front.matter.YamlFrontMatterExtension
import org.commonmark.ext.front.matter.YamlFrontMatterVisitor
import org.commonmark.parser.Parser
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isDirectory
import kotlin.io.path.isRegularFile
import kotlin.io.path.name
import kotlin.io.path.readText
import kotlin.io.path.relativeTo
import kotlin.io.path.walk

/**
 * Rule that requires skills to have proper frontmatter.
 *
 * @category skillFrontmatter
 */
object SkillFrontmatterRule : HarnessCheckRule() {
    /**
     * Category key.
     */
    override val category: String = "skillFrontmatter"

    private val markdownParser: Parser =
        Parser
            .builder()
            .extensions(listOf(YamlFrontMatterExtension.create()))
            .build()

    override fun validate(ctx: RuleContext): Collection<Finding> =
        buildList {
            val catObj = ctx.manifest.categoryObject(category)
            val parametersObj = catObj?.get("parameters")?.jsonObject
            if (catObj != null && parametersObj != null) {
                val dirPath = ctx.root / JsonAccess.stringFromObject(parametersObj, "rootDirectory")
                val severity = ctx.manifest.severityOf(category)
                when {
                    !dirPath.isDirectory() -> {
                        add(
                            Finding(
                                severity,
                                category,
                                ctx.manifest.stringValue(category, "missingDirectory").takeIf { message ->
                                    message.isNotEmpty()
                                } ?: ".claude/skills must contain at least one SKILL.md",
                            ),
                        )
                    }

                    else -> {
                        val files =
                            dirPath
                                .walk()
                                .filter { file -> file.isRegularFile() }
                                .filter { file -> file.name == JsonAccess.stringFromObject(parametersObj, "filename") }
                                .toList()
                        when {
                            files.isEmpty() -> {
                                add(
                                    Finding(
                                        severity,
                                        category,
                                        ctx.manifest.stringValue(category, "missingSkill").takeIf { message ->
                                            message.isNotEmpty()
                                        } ?: ".claude/skills must contain at least one SKILL.md",
                                    ),
                                )
                            }

                            else -> {
                                files.forEach { file ->
                                    val text = file.readText()
                                    when {
                                        !text.startsWith("---") -> {
                                            add(
                                                Finding(
                                                    severity,
                                                    category,
                                                    ctx.manifest
                                                        .stringValue(category, "missingFrontmatter")
                                                        .takeIf { message -> message.isNotEmpty() }
                                                        ?: "skill missing frontmatter: ${file.relativeTo(ctx.root)}",
                                                ),
                                            )
                                        }

                                        else -> {
                                            val frontmatter = extractFrontmatter(text)
                                            if (frontmatter["description"]?.firstOrNull()?.isNotBlank() != true) {
                                                add(
                                                    Finding(
                                                        severity,
                                                        category,
                                                        ctx.manifest
                                                            .stringValue(
                                                                category,
                                                                "missingDescription",
                                                            ).takeIf { message ->
                                                                message.isNotEmpty()
                                                            }
                                                            ?: "skill missing description: ${file.relativeTo(
                                                                ctx.root,
                                                            )}",
                                                    ),
                                                )
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

    private fun extractFrontmatter(content: String): Map<String, List<String>> =
        YamlFrontMatterVisitor()
            .also { visitor -> markdownParser.parse(content).accept(visitor) }
            .data
}
